src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -428,10 +429,21 @@ public static final class PlaceholderTask implements ContinuedTask, Serializable
         private Object readResolve() {
             if (cookie != null) {
                 synchronized (runningTasks) {
-                    runningTasks.put(cookie, new RunningTask());
+                    // If Jenkins stops while this step is resuming, there may be a PlaceholderTask in the queue as
+                    // well as in program.dat for the same step. We want to make sure not to create a second task to
+                    // avoid race conditions, so we use putIfAbsent.
+                    // TODO: This helps for runningTasks, but other fields like `stopping` may still be problematic.
+                    // Should we refactor things to guarantee that the relevant state is a singleton? For example,
+                    // introduce a PlaceholderTasksAction that holds a map of a new PlaceholderTaskState class, which
+                    // would hold most of what is currently in PlaceholderTask, and then PlaceholderTask would only
+                    // hold a `String cookie` field and would look up PlaceholderTaskState via the action so it wouldn't
+                    // matter where the task was serialized.
```

Comment:
IDK if there is a simpler option here. The problem is that in most cases `PlaceholderTask` is only serialized in `program.dat`, but whatever we do needs to be robust against deserialization from both the queue and `program.dat`. Perhaps the current approach is good enough but we could make `ExecutorStepDynamicContext.task` non-final and replace it with any task already in the queue if needed during resumption?

Reply:
> replace it with any task already in the queue

Why would a task already in queue not be `equals` to this one?

Reply:
They would be equal, but `PlaceholderTask` is not immutable (e.g. `stopping`), so it seems best to make sure that everything is using the same actual object regardless of code path.

Reply:
Ah maybe `stopping` can be moved to `PlaceholderExecutable` or `ExecutorStepExecution`?

Reply:
Perhaps, although now that I am looking I am wondering if you can also get weird issues if you restart while the initial `PlaceholderTask` is in the queue before `cookie` and `label` have been set to their final values.

Reply:
Right, this should perhaps be a different `Task` class for use after the initial scheduling.

Reply:
It seems like we have a few different ideas of how we could avoid problems here, but for now I am wondering if [42fedcb](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/344/commits/42fedcbe82f857bee30356e1504700b40c55263c) is good enough.

Reply:
Any thoughts on [42fedcb](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/344/commits/42fedcbe82f857bee30356e1504700b40c55263c)?

Reply:
Good enough for now.